File: iOS/Athlynk/Athlynk/Features/MacroHistoryView.swift
//
//  MacroHistoryView.swift
//  Past days of the MACRO food diary. Each day is editable: tap opens the
//  diary for that date (add/remove foods) via MacroLogView(logDate:).
//  Backed by /api/v1/nutrition/assignments/<id>/macro-history.
//

import SwiftUI

struct MacroHistoryView: View {
    let assignmentId: Int
    let planTitle: String

    @State private var days: [MacroHistoryDayDTO] = []
    @State private var hasMore = false
    @State private var loading = true
    @State private var loadingMore = false
    @State private var error: String?

    var body: some View {
        ZStack {
            VoltBackground(palette: [Palette.lime, Palette.cyan, Palette.amber, Palette.lime])
            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: 14) {
                    if loading {
                        MacroLogSkeleton()
                    } else if let error {
                        EmptyPanel(icon: "wifi.exclamationmark", text: error, color: Palette.lime)
                    } else if days.isEmpty {
                        EmptyPanel(icon: "calendar", text: "Nessun giorno registrato.\nInizia dal diario di oggi.",
                                   color: Palette.lime)
                    } else {
                        ForEach(Array(days.enumerated()), id: \.element.id) { i, day in
                            NavigationLink {
                                MacroLogView(assignmentId: assignmentId, planTitle: planTitle, logDate: day.date)
                            } label: { dayCard(day) }
                            .buttonStyle(PressableButtonStyle())
                            .revealUp(true, index: i)
                        }
                        if hasMore {
                            Button { Task { await loadMore() } } label: {
                                HStack(spacing: 8) {
                                    if loadingMore { ProgressView().tint(Palette.lime) }
                                    Text(loadingMore ? "Carico…" : "Carica altri giorni")
                                        .font(Typo.mono(12, .bold)).foregroundStyle(Palette.lime)
                                }
                                .frame(maxWidth: .infinity).padding(.vertical, 14)
                                .background(RoundedRectangle(cornerRadius: 12).fill(Palette.lime.opacity(0.12)))
                            }
                            .disabled(loadingMore)
                        }
                    }
                }
                .padding(.horizontal, 22).padding(.top, 12).padding(.bottom, 44)
            }
        }
        .navigationTitle("Storico pasti")
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
        .tint(Palette.lime)
        .task { await load() }
    }

    private func dayCard(_ day: MacroHistoryDayDTO) -> some View {
        let target = max(day.target.kcal, 1)
        let over = day.consumed.kcal > target
        let accent = over ? Palette.amber : Palette.lime
        return VStack(alignment: .leading, spacing: 12) {
            HStack(alignment: .firstTextBaseline) {
                Text(day.dowShort).font(Typo.mono(11, .black)).tracking(1.5).foregroundStyle(accent)
                Text(prettyDate(day.date)).font(Typo.display(17)).foregroundStyle(Palette.textHi)
                Spacer()
                Text("\(Int(day.consumed.kcal))").font(Typo.mono(17, .bold)).foregroundStyle(accent)
                Text("/ \(Int(day.target.kcal)) kcal").font(Typo.mono(10, .bold)).foregroundStyle(Palette.textMid)
            }
            HStack(spacing: 16) {
                miniMacro("P", day.consumed.protein, Palette.magenta)
                miniMacro("C", day.consumed.carb, Palette.cyan)
                miniMacro("G", day.consumed.fat, Palette.amber)
                Spacer()
                Text("\(day.entries.count) alimenti").font(Typo.mono(10)).foregroundStyle(Palette.textLow)
                Image(systemName: "chevron.right").font(.system(size: 12, weight: .black))
                    .foregroundStyle(accent.opacity(0.7))
            }
        }
        .padding(16).voltPanel(accent.opacity(0.3))
    }

    private func miniMacro(_ letter: String, _ grams: Double, _ color: Color) -> some View {
        HStack(spacing: 4) {
            Text(letter).font(Typo.mono(10, .black)).foregroundStyle(color)
            Text("\(Int(grams))g").font(Typo.mono(11, .semibold)).foregroundStyle(Palette.textMid)
        }
    }

    private func load() async {
        loading = true; error = nil
        do {
            let r = try await APIClient.shared.macroHistory(assignment: assignmentId)
            days = r.days; hasMore = r.hasMore
        } catch { self.error = error.localizedDescription }
        loading = false
    }

    private func loadMore() async {
        loadingMore = true
        do {
            let r = try await APIClient.shared.macroHistory(assignment: assignmentId, offset: days.count)
            days += r.days; hasMore = r.hasMore
        } catch { /* keep what we have */ }
        loadingMore = false
    }

    private func prettyDate(_ iso: String) -> String {
        let inF = DateFormatter(); inF.dateFormat = "yyyy-MM-dd"
        guard let d = inF.date(from: iso) else { return iso }
        let out = DateFormatter(); out.locale = Locale(identifier: "it_IT"); out.dateFormat = "d MMM yyyy"
        return out.string(from: d)
    }
}

File: android/lib/features/athlete/nutrition/macro_history_view.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/api/athlete_api.dart';
import '../../../core/l10n/formatters.dart';
import '../../../core/models/models.dart';
import '../../../core/providers.dart';
import '../../../design/components/misc.dart';
import '../../../design/components/panel.dart';
import '../../../design/components/scaffold.dart';
import '../../../design/components/skeleton.dart';
import '../../../design/theme.dart';
import 'macro_log_view.dart';

/// Past MACRO days — port of iOS `MacroHistoryView`, 14 per page; tapping a
/// day opens the (editable) diary for that date.
class MacroHistoryView extends ConsumerStatefulWidget {
  const MacroHistoryView({super.key, required this.assignmentId});

  final int assignmentId;

  @override
  ConsumerState<MacroHistoryView> createState() => _MacroHistoryViewState();
}

class _MacroHistoryViewState extends ConsumerState<MacroHistoryView> {
  List<MacroHistoryDayDto>? _days;
  bool _error = false;
  bool _hasMore = false;
  bool _loadingMore = false;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    try {
      final res =
          await ref.read(apiClientProvider).macroHistory(widget.assignmentId);
      if (mounted) {
        setState(() {
          _days = res.days;
          _hasMore = res.hasMore;
        });
      }
    } catch (_) {
      if (mounted && _days == null) setState(() => _error = true);
    }
  }

  Future<void> _loadMore() async {
    if (_loadingMore) return;
    setState(() => _loadingMore = true);
    try {
      final res = await ref
          .read(apiClientProvider)
          .macroHistory(widget.assignmentId, offset: _days?.length ?? 0);
      if (mounted) {
        setState(() {
          _days = [...?_days, ...res.days];
          _hasMore = res.hasMore;
        });
      }
    } catch (_) {} finally {
      if (mounted) setState(() => _loadingMore = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final days = _days;
    return Scaffold(
      backgroundColor: Colors.transparent,
      appBar: AppBar(),
      body: ScreenScroll(
        topPadding: 0,
        spacing: Space.element,
        onRefresh: _load,
        children: [
          const ScreenHeader(eyebrow: 'Diario alimentare', title: 'Storico'),
          if (days == null && !_error)
            const DateCardsSkeleton()
          else if (_error)
            EmptyPanel.network(onCta: () {
              setState(() => _error = false);
              _load();
            })
          else if (days!.isEmpty)
            const EmptyPanel(
              icon: Icons.calendar_view_day_outlined,
              message: 'Nessun giorno registrato finora.',
            )
          else ...[
            for (final day in days) _dayRow(day),
            if (_hasMore)
              LoadMoreButton(onTap: _loadMore, loading: _loadingMore),
          ],
        ],
      ),
    );
  }

  Widget _dayRow(MacroHistoryDayDto day) {
    final date = Formatters.parseDate(day.date);
    final over = day.consumed.kcal > day.target.kcal && day.target.kcal > 0;
    return VoltPanel(
      onTap: date == null
          ? null
          : () => Navigator.of(context).push(MaterialPageRoute<void>(
                builder: (_) => MacroLogView(
                  assignmentId: widget.assignmentId,
                  logDate: date,
                ),
              )),
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
      child: Row(
        children: [
          SizedBox(
            width: 46,
            child: Column(
              children: [
                Text(day.dowShort,
                    style: Typo.mono(9, FontWeight.w700, Palette.cyan)),
                Text(date == null ? '—' : '${date.day}',
                    style: Typo.poster(20)),
              ],
            ),
          ),
          const SizedBox(width: 10),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  '${day.consumed.kcal.toInt()} / ${day.target.kcal.toInt()} kcal',
                  style: Typo.body(14.5, FontWeight.w700,
                      over ? Palette.amber : Palette.textHi),
                ),
                Text(
                  'P ${day.consumed.protein.toInt()} · C ${day.consumed.carb.toInt()} · F ${day.consumed.fat.toInt()} · ${day.entries.length} alimenti',
                  style: Typo.mono(9.5, FontWeight.w600, Palette.textMid),
                ),
              ],
            ),
          ),
          const Icon(Icons.chevron_right_rounded,
              size: 17, color: Palette.textLow),
        ],
      ),
    );
  }
}
